mirrord-layer/src/file.rs

```diff
@@ -0,0 +1,170 @@
+        r".*\.so",
+        r".*\.d",
+        r"^/proc/.*",
+        r"^/sys/.*",
+        r"^/lib/.*",
+        r"^/etc/.*",
+        r"^/usr/.*",
+        r"^/dev/.*",
+        // TODO(alex) [low] 2022-05-13: Investigate why it's trying to load this file all of a
+        // sudden. We have to ignore this here as `node` will look for it outside the cwd (it
+        // searches up the directory, until it finds).
+        r".*/package.json",
+        cwd_str,
+    ])
+    .unwrap();
+
+    set
+});
+
+/// Blocking wrapper around `libc::open` call.
+///
+/// It's bypassed when trying to load system files, and files from the current working directory
+/// (which is different anyways when running in `-agent` context).
+///
+/// When called for a valid file, it'll block, send a `ClientMessage::OpenFileRequest` to be handled
+/// by `mirrord-agent` (`handle_peer_message` function), and wait until it receives a
+/// `DaemonMessage::FileOpenResponse`.
+pub fn open(raw_path: *const c_char, oflag: c_int) -> RawFd {
+    let path: PathBuf = unsafe { CStr::from_ptr(raw_path) }
+        .to_str()
+        .expect("Failed converting path from c_char!")
+        .into();
+
+    if IGNORE_FILES.is_match(path.to_str().unwrap()) {
+        let fd = unsafe { libc::open(raw_path, oflag) };
+        fd
+    } else {
+        debug!("open -> trying to open valid file {path:?}");
+        let sender = unsafe { HOOK_SENDER.as_ref().unwrap() };
+        let (file_channel_tx, file_channel_rx) = oneshot::channel::<FileOpenResponse>();
+
+        let requesting_file = OpenFileHook {
+            path,
+            file_channel_tx,
+        };
+
+        match sender.blocking_send(HookMessage::OpenFileHook(requesting_file)) {
+            Ok(_) => {}
+            Err(fail) => {
+                error!("open: failed to send open message: {fail:?}!");
+                return libc::EFAULT;
+            }
+        };
+
+        debug!("Blocking on file open operation!");
+
+        let file_open = RUNTIME.block_on(async {
+            let file_open = file_channel_rx.await.unwrap();
+            file_open
+        });
+
+        debug!("After `block_on` call, we have an open file {file_open:#?}!");
+
+        // TODO(alex) [high] 2022-05-12: Instead of returning this `fake_fd` thing, we should block
+        // while waiting for an `-agent` reply.
+        // To do this:
+        // - create a (send_tx, recv_tx) for file ops;
+        // - do `recv_tx.recv()` here to block until the "file is open" message is received;
+        // - send an open file request to `-agent` in `poll_agent`;
+        // - when the reply comes back, `send_tx.send(file)` in `poll_agent`;
+
+        file_open.fd
+    }
+}
+
+/// NOTE(alex): libc also has an `open64` function. Both functions point to the same address, so
+/// trying to intercept them all will result in an `InterceptorAlreadyReplaced` error.
+pub(super) unsafe extern "C" fn open_detour(path: *const c_char, oflag: c_int) -> RawFd {
```

Comment:
I would consider only hooking read operations atm? @eyalb181 UX wise it might be better initially